shell.go

```diff
@@ -121,6 +126,39 @@ func NewShellWithClient(url string, client *gohttp.Client) *Shell {
 	return &sh
 }
 
+// encodedAbsolutePathVersion is the version from which the absolute path header in
+// multipart requests is %-encoded. Before this version, its sent raw.
+var encodedAbsolutePathVersion = semver.MustParse("0.23.0-dev")
+
+func (s *Shell) loadRemoteVersion() error {
+	if s.version == nil {
+		version, _, err := s.Version()
+		if err != nil {
+			return err
+		}
+
+		remoteVersion, err := semver.New(version)
+		if err != nil {
+			return err
+		}
+
+		s.versionOnce.Do(func() {
+			s.version = remoteVersion
+		})
+	}
+
+	return nil
+}
```

Comment:
I'm pretty sure the `.Do` should englobe all of loadRemoteVersion's body.
Because this code will concurrently request the version number but save it only once, this leads to a data race between your `s.version == nil` and `s.version = remoteVersion`.